AI/Nullkiller2/Engine/ResourceTrader.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+* ResourceTrader.cpp, part of VCMI engine
+*
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*/
+#include "StdInc.h"
+#include "ResourceTrader.h"
+
+namespace NK2AI
+{
+bool ResourceTrader::trade(BuildAnalyzer & buildAnalyzer, CCallback & cc, const TResources & freeResources)
+{
+	bool haveTraded = false;
+	ObjectInstanceID marketId;
+
+	// TODO: Mircea: What about outside town markets that have better rates than a single town for example?
+	// Are those used anywhere? To inspect.
+	for(const auto * const town : cc.getTownsInfo())
+	{
+		if(town->hasBuiltSomeTradeBuilding())
```

Comment:
Huh, interesting. Your code is actually incorrect and can pick town without marketplace if it has some other trade building, like Freelancer Guild. 

But what is broken is `hasBuiltSomeTradeBuilding` method - it should check for presence of `EMarketMode::RESOURCE_RESOURCE`, not for *any* mode. All users of this method expect such logic.